ethcore/light/src/on_demand/mod.rs

```diff
@@ -435,82 +501,56 @@ impl OnDemand {
 	// dispatch pending requests, and discard those for which the corresponding
 	// receiver has been dropped.
 	fn dispatch_pending(&self, ctx: &BasicContext) {
-		if self.pending.read().is_empty() { return }
-		let mut pending = self.pending.write();
+		if self.pending.read().is_empty() {
+			return
+		}
 
-		debug!(target: "on_demand", "Attempting to dispatch {} pending requests", pending.len());
+		let mut pending = self.pending.write();
 
 		// iterate over all pending requests, and check them for hang-up.
 		// then, try and find a peer who can serve it.
 		let peers = self.peers.read();
-		*pending = ::std::mem::replace(&mut *pending, Vec::new()).into_iter()
+
+		*pending = ::std::mem::replace(&mut *pending, Vec::new())
+			.into_iter()
 			.filter(|pending| !pending.sender.is_canceled())
 			.filter_map(|mut pending| {
-				// the peer we dispatch to is chosen randomly
+
 				let num_peers = peers.len();
-				let history_len = pending.query_id_history.len();
-				let offset = if history_len == 0 {
-					pending.remaining_query_count = self.base_retry_count;
-					let rand = rand::random::<usize>();
-					pending.base_query_index = rand;
-					rand
-				} else {
-					pending.base_query_index + history_len
-				} % cmp::max(num_peers, 1);
-				let init_remaining_query_count = pending.remaining_query_count; // to fail in case of big reduction of nb of peers
-				for (peer_id, peer) in peers.iter().chain(peers.iter())
-					.skip(offset).take(num_peers) {
-					// TODO: see which requests can be answered by the cache?
-					if pending.remaining_query_count == 0 {
-						break
+				// The first peer to dispatch the request is chosen at random
+				let rand = rand::random::<usize>() % cmp::max(num_peers, 1);
```

Comment:
Can we use here `gen_range` instead?
```suggestion
				let rand = rand::thread_rng().gen_range(0, cmp::max(num_peers, 1));
```

Reply:
Thanks, I rushed that part but it is fixed now at least!